pulsar-client-cpp/lib/auth/AuthOauth2.cc

```diff
@@ -143,6 +143,18 @@ ClientCredentialFlow::ClientCredentialFlow(ParamMap& params)
       audience_(params["audience"]),
       scope_(params["scope"]) {}
 
+const std::string ClientCredentialFlow::getWellKnownOpenIdConfigurationUrl() {
```

Comment:
```suggestion
std::string ClientCredentialFlow::getWellKnownOpenIdConfigurationUrl() const {
```

Reply:
I want the return value not to be changed. Maybe we don't need the const qualifier here?

Reply:
In C++, returning a const value is meaningless, for example, given the following function that returns a const `std::string`

```c++
const std::string f() { return "hello"; }
```

Then we have:

```c++
// It's an assignment, i.e. deep copy in Java, so s1 is non-const
auto s1 = f();
s1.append("world"); // OK because the type of s1 is non-const
// s2 is a const reference, but it's dangerous in C++ because f() returns a temporary variable.
auto& s2 = f();
// s2.append("world"); // Failed because s2 references to a const object
```

It's dangerous to use `s2` in the above example because it references a temporary variable. See following example:

```c++
#include <iostream>
using namespace std;

const std::string f() { return "hello"; }

const std::string& g() {
    auto& s = f();
    return s;
}

int main(int argc, char* argv[]) {
    auto& s = g();
    cout << s << endl;
    return 0;
}
```

When you compile this program, you can see the warning from the Clang output:

```
1.cc:8:12: warning: returning reference to local temporary object [-Wreturn-stack-address]
    return s;
           ^
1.cc:7:11: note: binding reference variable 's' here
    auto& s = f();
```

And the output is undefined.

Reply:
BTW, usually, we should mark all methods that don't change the internal state as `const`. For example,

```c++
struct A {
    int x;

    int getX() {
        x++;  // a getter should not modify the internal state, but the compilation can pass
        return x;
    }
};
```

In this case, you should make `getX` a `const` method like:

```c++
    int getX() const {
        // x++;  // Compilation error! Because it changes the value of the field of A
        return 2;
    }
```

The const qualifier makes your code clear (because we can know which methods don't change the state) and prevent unexpected changes (for example, modifying the state in a getter).

See https://google.github.io/styleguide/cppguide.html#Use_of_const

Reply:
Thanks for your explanation, I'll fix this.